src/Plugin/Field/FieldFormatter/PlayerFormatter.php

```diff
@@ -0,0 +1,194 @@
+
+  /**
+   * The creator used to load player factories.
+   *
+   * @var \Drupal\media_mpx\DataObjectFactoryCreator
+   */
+  private $dataObjectFactoryCreator;
+
+  /**
+   * Constructs a PlayerFormatter object.
+   *
+   * @param string $plugin_id
+   *   The plugin_id for the formatter.
+   * @param mixed $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Drupal\Core\Field\FieldDefinitionInterface $field_definition
+   *   The definition of the field to which the formatter is associated.
+   * @param array $settings
+   *   The formatter settings.
+   * @param string $label
+   *   The formatter label display setting.
+   * @param string $view_mode
+   *   The view mode.
+   * @param array $third_party_settings
+   *   Any third party settings.
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
+   *   The entity type manager.
+   * @param \Drupal\media_mpx\DataObjectFactoryCreator $data_object_factory_creator
+   *   The creator of mpx data factories.
+   */
+  public function __construct(string $plugin_id, $plugin_definition, FieldDefinitionInterface $field_definition, array $settings, string $label, string $view_mode, array $third_party_settings, EntityTypeManagerInterface $entity_type_manager, DataObjectFactoryCreator $data_object_factory_creator) {
+    parent::__construct($plugin_id, $plugin_definition, $field_definition, $settings, $label, $view_mode, $third_party_settings);
+    $this->entityTypeManager = $entity_type_manager;
+    $this->dataObjectFactoryCreator = $data_object_factory_creator;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $plugin_id,
+      $plugin_definition,
+      $configuration['field_definition'],
+      $configuration['settings'],
+      $configuration['label'],
+      $configuration['view_mode'],
+      $configuration['third_party_settings'],
+      $container->get('entity_type.manager'),
+      $container->get('media_mpx.data_object_factory_creator')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function viewElements(FieldItemListInterface $items, $langcode) {
+    $element = [];
+
+    /** @var \Drupal\media\Entity\Media $entity */
+    $entity = $items->getEntity();
+    /** @var \Drupal\media_mpx\Plugin\media\Source\Media $source_plugin */
+    $source_plugin = $entity->getSource();
+
+    $factory = $this->dataObjectFactoryCreator->forObjectType($source_plugin->getAccount()->getUserEntity(), 'Player Data Service', 'Player', '1.6');
+    $player = $factory->load($this->getSetting('player'))->wait();
+
+    foreach ($items as $delta => $item) {
+      /** @var \Lullabot\Mpx\DataService\Media\Media $mpx_media */
+      $mpx_media = $source_plugin->getMpxObject($entity);
+      $url = new Url($source_plugin->getAccount(), $player, $mpx_media);
+
+      // What cache contexts or tags do we set?
```

Comment:
Can't we define custom tags for mpx clips? Then, if a clip is updated, we would invalidate them.

Reply:
I'm not sure if we need to define anything custom - since they are media entities, the tags should be available. I'm just not aware if the FAPI adds them automatically (probably does, just want to be sure!). I will label this as `@todo` so it shows up.